### PR TITLE
Backport 2.16: Test read of zero-padded negative number

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
This is a partial backport of https://github.com/ARMmbed/mbedtls/pull/4297.

* Backport the added unit tests for `mbedtls_mpi_read_string`.
* This revealed that `"-0"` is parsed as a negative 0, which is not handled correctly everywhere. (`mbedtls_mpi_cmp_mpi` does say that +0 and -0 are equal, but `mbedtls_mpi_lt_mpi_ct` doesn't.) So backport the change to `mbedtls_mpi_read_string` as well, even if it's only for the sake of this edge case (the original motivation for this change, which is that negative numbers with a leading zero were parsed incorrectly, does not apply to 2.16.).  
    * Code size impact: saves a few bytes (15625 → 15605 for our reference M0+ build with my version of GCC).
* Backport the added unit tests for `mbedtls_ecp_muladd`.
* Do not backport the bug fix to `fix_negative` since it was only briefly broken in `development.
* Do not backport the unit tests for `fix_negative` because it's an internal function and it's too much work to expose it in 2.16 (doable, but I don't think it's worth the effort).

### CODE REVIEW

File: tests/suites/test_suite_mpi.data [hunk 1]
```diff
@@ -10,21 +10,39 @@ mpi_null:
...
+mpi_read_write_string:16:"-0":16:"00":100:0:0
+
 Base test mpi_read_write_string #3 (Negative decimal)
 mpi_read_write_string:10:"-23":10:"-23":100:0:0
 
-Base test mpi_read_write_string #3 (Negative hex)
+Base test mpi_read_write_string #3 (Negative decimal, leading 0)
+mpi_read_write_string:10:"-023":10:"-23":100:0:0
+
+Base test mpi_read_write_string #3 (Negative decimal -> hex)
```
Comment (mpg) [inline tests/suites/test_suite_mpi.data:37]: That should be hex -> decimal

**Review by mpg — CHANGES_REQUESTED** [review]
Looks almost entirely good to me.

I agree with the PR description about what's relevant to backport here and what's not relevant or not worth it. I checked that each commit here is a proper backport of a corresponding commit in #4297 or #4323.

The only thing is that one test description is wrong, and apparently the commit that fixed it in the original PR was not cherry-picked here.

**Review by mpg — APPROVED** [review]
Thanks for fixing the typo. Looks all good to me now.

**Review by chris-jones-arm — APPROVED** [review]
I agree with the scope of the backport and I have verified that it is faithful to the original.

**Pull request merged**